// @falloutstudios/djs-pagination/dist/lib/types/enums.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PaginationControllerType = exports.SendAs = exports.getEnumValue = void 0;
function getEnumValue(enum_, key) {
    // @ts-expect-error
    return typeof key === 'string' ? enum_[key] : key;
}
exports.getEnumValue = getEnumValue;
var SendAs;
(function (SendAs) {
    SendAs[SendAs["NewMessage"] = 1] = "NewMessage";
    SendAs[SendAs["EditMessage"] = 2] = "EditMessage";
    SendAs[SendAs["ReplyMessage"] = 3] = "ReplyMessage";
})(SendAs = exports.SendAs || (exports.SendAs = {}));
var PaginationControllerType;
(function (PaginationControllerType) {
    PaginationControllerType[PaginationControllerType["FirstPage"] = 1] = "FirstPage";
    PaginationControllerType[PaginationControllerType["PreviousPage"] = 2] = "PreviousPage";
    PaginationControllerType[PaginationControllerType["NextPage"] = 3] = "NextPage";
    PaginationControllerType[PaginationControllerType["LastPage"] = 4] = "LastPage";
    PaginationControllerType[PaginationControllerType["Stop"] = 5] = "Stop";
})(PaginationControllerType = exports.PaginationControllerType || (exports.PaginationControllerType = {}));
